mountpoint-s3/src/data_cache/disk_data_cache.rs

```diff
@@ -0,0 +1,315 @@
+//! Module for the on-disk data cache implementation.
+
+use std::fs;
+use std::io::{ErrorKind, Write};
+use std::ops::RangeBounds;
+use std::path::PathBuf;
+
+use base64ct::{Base64UrlUnpadded, Encoding};
+use bytes::{BufMut, BytesMut};
+use mountpoint_s3_crt::checksums::crc32c;
+use tracing::{error, trace, warn};
+
+use crate::data_cache::DataCacheError;
+
+use super::{BlockIndex, CacheKey, ChecksummedBytes, DataCache, DataCacheResult};
+
+/// On-disk implementation of [DataCache].
+///
+/// TODO: Store checksums on disk, reconstruct as [ChecksummedBytes] using same checksum avoiding recomputation.
+///
+/// TODO: Store additional metadata with each block such as expected S3 key, ETag, etc..
+pub struct DiskDataCache {
+    block_size: u64,
+    cache_directory: PathBuf,
+}
+
+impl DiskDataCache {
+    /// Create a new instance of an [DiskDataCache] with the specified `block_size`.
+    pub fn new(cache_directory: PathBuf, block_size: u64) -> Self {
+        DiskDataCache {
+            block_size,
+            cache_directory,
+        }
+    }
+
+    /// Get the relative path for the given block.
+    fn get_path_for_key(&self, cache_key: &CacheKey) -> PathBuf {
+        let mut path = self.cache_directory.join("v1");
+
+        // An S3 key may be up to 1024 UTF-8 bytes long, which exceeds the maximum UNIX file name length.
+        // Instead, we encode the key and split into 255 character long directory names.
+        let encoded_s3_key = Base64UrlUnpadded::encode_string(cache_key.s3_key.as_bytes());
+        let mut slice = encoded_s3_key.as_str();
+        while !slice.is_empty() {
+            let (chunk, remaining) = slice.split_at(255.min(slice.len()));
+            path.push(chunk);
+            slice = remaining;
+        }
+
+        path.push(cache_key.etag.as_str());
+        path
```

Comment:
As mentioned elsewhere: think this should just be a SHA256 hash of the object key, etag, and block index. It gets us out of all the concerns about filename size and valid/invalid characters, and obfuscates the cache so customers don't take a dependency on its layout (without a lot of effort). There's always a collision risk with hashes but these are big enough that I wouldn't worry about it, and since you're eventually going to include the key and etag in the file contents we can always disambiguate.

Reply:
I had previously preferred to keep the ETag and Block Index exposed.

My reasoning here was that:
- We can wipe all the blocks for a whole file version relatively easily.
- When we know that an object has been replaced, we could purge all cache versions by wiping all blocks that don't match the new ETag.

However, we still plan to use LRU cache eviction and those versions would eventually be purged. The bit that makes me uncomfortable is that we would be allowing the disk to be consumed unnecessarily in those use cases by waiting for them to be evicted when we run out of capacity. Possible alternatives might be that we have a TTL on the data cache entries also.

Reply:
I am coming around to the idea of using the SHA256 over Base64. Agree that it should be very rare, and if we do get a collision then we will check the S3 key encoded in the block and reject blocks that don't match. Small concern over thrashing with blocks being written over each other, but again: we think this is rare.

If we do move to SHA256 (or similar) then I'll add some metrics for collisions.

Reply:
What scenarios would we need to evict more than one old version of an object from the cache? My thinking is that at any point in time there's at most one cached version of an object — when we discover the remote etag has changed, we just evict whatever the current version is (whose etag we know since it's already in the inode).

Even if it's more complicated than that, I don't think it's worth the trouble to try to accelerate eviction right now; seems fine to just fall back on those objects eventually aging out of the LRU.

Reply:
OK yeah, sounds good. Agree, I think we should have at most one version in the cache. Maybe the odd block if doing something weird where two file handles have two objects open, but should be cleaned up by LRU.

Reply:
I'll keep the ETag and Block ID in the path for now just in the interest of getting a new version of this PR out, and we can simplify in another PR.

I will move to SHA256 in this PR though for the S3 key.

Reply:
Opened: #593